src/main/java/org/truffleruby/extra/ffi/Pointer.java

```diff
@@ -0,0 +1,187 @@
+
+    public static final Pointer NULL = new Pointer(0);
+
+    public static Pointer malloc(long size) {
+        return new Pointer(UNSAFE.allocateMemory(size));
+    }
+
+    private final long address;
+    private boolean autorelease;
+
+    public Pointer(long address) {
+        this.address = address;
+    }
+
+    public void writeByte(long offset, byte b) {
+        UNSAFE.putByte(address + offset, b);
+    }
+
+    public void writeShort(long offset, short value) {
+        UNSAFE.putShort(address + offset, value);
+    }
+
+    public void writeInt(long offset, int value) {
+        UNSAFE.putInt(address + offset, value);
+    }
+
+    public void writeLong(long offset, long value) {
+        UNSAFE.putLong(address + offset, value);
+    }
+
+    public void writePointer(long offset, Pointer value) {
+        writeLong(offset, value.getAddress());
+    }
+
+    private void writeZeroTerminatedBytes(long offset, byte[] bytes, int start, int length) {
+        writeBytes(offset, bytes, start, length);
+        writeByte(offset + length, (byte) 0);
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public void writeString(long offset, String string, int maxLength, Charset cs) {
+        ByteBuffer buf = cs.encode(string);
+        int len = Math.min(maxLength, buf.remaining());
+        writeZeroTerminatedBytes(offset, buf.array(), buf.arrayOffset() + buf.position(), len);
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public void writeBytes(long offset, long size, byte value) {
+        for (long n = 0; n < size; n++) {
+            writeByte(offset + n, value);
+        }
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public void writeBytes(long offset, byte[] bytes, int index, int length) {
+        for (int n = 0; n < length; n++) {
+            writeByte(offset + n, bytes[index + n]);
+        }
+    }
+
+    public byte readByte(long offset) {
+        return UNSAFE.getByte(address + offset);
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public void readBytes(long from, byte[] buffer, int bufferPos, int i) {
+        for (int n = 0; n < i; n++) {
+            buffer[bufferPos + n] = readByte(from + n);
+        }
+    }
+
+    public short readShort(long offset) {
+        return UNSAFE.getShort(address + offset);
+    }
+
+    public int readInt(long offset) {
+        return UNSAFE.getInt(address + offset);
+    }
+
+    public long readLong(long offset) {
+        return UNSAFE.getLong(address + offset);
+    }
+
+    private byte[] readZeroTerminatedByteArray(long offset) {
+        final int length = findByte(offset, (byte) 0);
+        final byte[] bytes = new byte[length];
+        readBytes(offset, bytes, 0, length);
+        return bytes;
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public String readString(long offset) {
+        return Charset.defaultCharset().decode(ByteBuffer.wrap(readZeroTerminatedByteArray(offset))).toString();
```

Comment:
Yes I don't quite understand why this is the way it is, but this is the behaviour of the old JNR read string, which seems to do what we want, so I didn't want to tweak it here.